### PR TITLE
IUrlHelper and support for RouteOptions.LowercaseUrls

### PR DESCRIPTION
**Pull request by kishanAnem**
IUrlHelper and support for RouteOptions.LowercaseUrls

#issue : https://github.com/aspnet/Mvc/issues/7720

### CODE REVIEW

**Comment by rynowak** [comment]
From the issue:

> if you'd like to submit a PR in the Routing repo to handle this with **a separate option** we'd be something we'd consider to take.

This needs to be a separate option or else it's a breaking change.

**Comment by kishan3034** [comment]
@rynowak this is a enhancement. it is not breaking anything.
expected result: query?property=x
actual result: query?propertySomething=x

**Comment by kishan3034** [comment]
hi @rynowak,

please guide me.im interested to work.

**Comment by rynowak** [comment]
Start be adding another property to `RouteOptions` - `bool LowercaseQueryStrings` and then make the new behavior respect that setting. Also you should update the documentation for `LowercaseUrls` to point out that it doesn't include the query string.

**Comment by kishan3034** [comment]
thanks @rynowak.
i’ll do @rynowak.

**Comment by kishanAnem** [comment]
Hi @rynowak,
I made changes. Please review.

File: src/Microsoft.AspNetCore.Routing/RouteCollection.cs [hunk 1]
```diff
@@ -138,9 +138,9 @@ private VirtualPathData NormalizeVirtualPath(VirtualPathData pathData)
 
             var url = pathData.VirtualPath;
 
-            if (!string.IsNullOrEmpty(url) && (_options.LowercaseUrls || _options.AppendTrailingSlash))
+            if (!string.IsNullOrEmpty(url) && (_options.LowercaseUrls || _options.LowercaseQueryStrings || _options.AppendTrailingSlash))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/RouteCollection.cs]: `LowercaseQueryStrings` shouldn't appear here, that just applies to the query string.

File: src/Microsoft.AspNetCore.Routing/RouteOptions.cs [hunk 2]
```diff
@@ -10,10 +10,15 @@ namespace Microsoft.AspNetCore.Routing
     public class RouteOptions
     {
         /// <summary>
-        /// Gets or sets a value indicating whether all generated URLs are lower-case.
+        /// Gets or sets a value indicating whether all generated URLs are lower-case. it doesn't include the query string.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/RouteOptions.cs]: Suggestions for these:

`Gets or sets a value indicating whether all generated paths URLs are lower-case. Use <see cref="LowercaseQueryStrings" /> to configure the behavior for query strings.`

File: src/Microsoft.AspNetCore.Routing/RouteOptions.cs [hunk 3]
```diff
@@ -10,10 +10,15 @@ namespace Microsoft.AspNetCore.Routing
...
         /// </summary>
         public bool LowercaseUrls { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether all generated QUERY STRINGS are lower-case. property will when LowercaseUrls true.
+        /// </summary>
+        public bool LowercaseQueryStrings { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/RouteOptions.cs]: Suggesting for this:
`Gets or sets a value indicating whether a generated query strings are lower-case. This property will be unless <see cref="LowercaseUrls" /> is also <c>true</c>`

File: test/Microsoft.AspNetCore.Routing.Tests/RouteCollectionTest.cs [hunk 4]
```diff
@@ -682,12 +701,13 @@ private static INamedRouter CreateNamedRoute(string name, bool accept = false, s
 
         private static Action<RouteOptions> GetRouteOptions(
             bool lowerCaseUrls = false,
-            bool appendTrailingSlash = false)
+            bool appendTrailingSlash = false, bool lowercaseQueryStrings = false)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/RouteCollectionTest.cs]: please put the new parameter on it's own line. This makes diffs nicer in the future.

**Comment by kishanAnem** [comment]
Hi @rynowa,
Thank you very much for the support. 
You are guiding me step by step.

**Pull request merged**